proof_ipa.go

```diff
@@ -132,3 +133,66 @@ func SerializeProof(proof *Proof) ([]byte, error) {
 
 	return buf.Bytes(), nil
 }
+
+func DeserializeProof(proofSerialized []byte) (*Proof, error) {
+	var (
+		numPoaStems, numExtStatus, numCommitments uint32
+		poaStems [][]byte
+		extStatus []byte
+		commitments []*Point
+		multipoint ipa.MultiProof
+	)
+	reader := bytes.NewReader(proofSerialized)
+
+	if err := binary.Read(reader, binary.LittleEndian, &numPoaStems); err != nil {
+		return nil, err
+	}
+	poaStems = make([][]byte, numPoaStems)
+	for i := 0; i < int(numPoaStems); i++ {
+		var poaStem [31]byte
+		if err := binary.Read(reader, binary.LittleEndian, &poaStem); err != nil {
+			return nil, err
+		}
+
+		poaStems[i] = poaStem[:]
+	}
```

Comment:
This allocate a buffer and copies the data into it, when you already have direct access to the source data.

```suggestion
	// verify that there is enough data to read
	if numPoaStems * 31 + 4 <= len(proofSerialized) {
		return nil, errors.New("truncated proof")
	}
	poaStems = make([][]byte, numPoaStems)
	for i := 0; i < int(numPoaStems); i++ {
		poaStems[i] = poaStem[4+31*i:4+31*i+31]
		reader.Seek(31, io.SeekStart)
	}
```

Reply:
Okay.  If we go with this option, then the caller must make sure that the source data is not used after they call `DeserializeProof(sourceData)`.  It has potential to create some annoying bugs down the line.

But I'm fine to make it this way if you think the reduction in allocations is worth it here.

Reply:
It's only a problem if the buffer is reused to deserialize another proof afterwards, it's unlikely to happen but I guess you are right. I made a note of this in issue #176 and we will revisit it when the go-ethereum part of the code is more stable, and we know that this won't happen.